compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -582,6 +596,68 @@ fn report_missing_placeholders(
         })
     };
 
+    let placeholders = pieces
+        .iter()
+        .filter_map(|piece| {
+            if let parse::Piece::NextArgument(argument) = piece && let ArgumentNamed(binding) = argument.position {
+                let span = fmt_span.from_inner(InnerSpan::new(argument.position_span.start, argument.position_span.end));
+                Some((span, binding))
+            } else { None }
+        })
+        .collect::<Vec<_>>();
+
+    let mut args_spans = vec![];
+    let mut fmt_spans = FxIndexSet::default();
+
+    for (i, unnamed_arg) in args.unnamed_args().iter().enumerate().rev() {
+        let Some(ty) = unnamed_arg.expr.to_ty() else { continue };
+        let Some(argument_binding) = ty.kind.is_simple_path() else { continue };
+        let argument_binding = argument_binding.as_str();
+
+        if used[i] {
+            continue;
+        }
+
+        let matching_placeholders = placeholders
+            .iter()
+            .filter(|(_, inline_binding)| argument_binding == *inline_binding)
+            .collect::<Vec<_>>();
+
+        if !matching_placeholders.is_empty() {
+            args_spans.push(unnamed_arg.expr.span);
+            for placeholder in &matching_placeholders {
+                fmt_spans.insert(*placeholder);
+            }
+        }
+    }
+
+    if !args_spans.is_empty() {
+        let mut multispan = MultiSpan::from(args_spans.clone());
+
+        let msg = if fmt_spans.len() > 1 {
+            "the formatting strings already captures the bindings \
+            directly, they don't need to be included in the argument list"
+        } else {
+            "the formatting string already captures the binding \
+            directly, it doesn't need to be included in the argument list"
+        };
+
+        for (span, binding) in fmt_spans {
+            multispan.push_span_label(
+                *span,
+                format!("this formatting specifier is referencing the `{binding}` binding"),
+            );
+        }
+
+        for span in &args_spans {
+            multispan.push_span_label(*span, "this can be removed");
```

Comment:
It would be great if this is were a suggestion so that it could be automatically fixed.

Reply:
When working on this I decided to go for the "conservative" option because in some cases (like the one below) it's not entirely clear what the user meant:

```
help: the formatting string already captures the binding directly, it doesn't need to be included in the argument list
  --> tests/ui/did_you_mean/issue-105225.rs:17:43
   |
17 |     println!("{} {} {x} {y} {}", x, x, x, y, y);
   |                          -                ^  ^ this can be removed
   |                          |                |
   |                          |                this can be removed
   |                          this formatting specifier is referencing the `y` binding
```

Maybe we could only suggest it when the intent is clear?

Reply:
Maybe we could only suggest it when the intent is clear?

👍